srcjs/input_binding_text.js

```diff
@@ -27,8 +27,27 @@ $.extend(textInputBinding, {
     if (data.hasOwnProperty('value'))
       this.setValue(el, data.value);
 
-    if (data.hasOwnProperty('label'))
-      $(el).parent().find('label[for="' + $escape(el.id) + '"]').text(data.label);
+
+    var labelTag = $(el).parent().find('label[for="' + $escape(el.id) + '"]');
+    var hasLabelTag = labelTag.length > 0;
+
+    // If data.label exists, then we may need to insert a label
+    // tag into the DOM. If it doesn't exist, then the label tag
+    // should be removed to be more consistent with the behavior
+    // of `textInput(label = NULL)`
+    if (data.hasOwnProperty('label')) {
+
+      if (hasLabelTag) {
+        labelTag.text(data.label);
+      } else {
+        $('<label for="' + $escape(el.id) + '">' + data.label + '</label>').insertBefore(el);
+      }
+
+    } else {
```

Comment:
I don't think it will ever get to this `else` because in `updateTextInput`, it drops `NULL`s:

https://github.com/rstudio/shiny/blob/386078d4415da7bbb05b8d63874758d5f6b1672f/R/update-input.R#L39

Reply:
Oh actually, I was incorrect. It will encounter this code path, but it will get there even if the user doesn't mean to clear the label. If they call `updateTextInput(session, "foo", value = "bar")`, then it will clear the label.

Reply:
Have a look at how `selectInput()` handles this with `character(0)`